File: cases/taint/scala/context-pair-depth2-negative/ContextPairDepth2Negative.scala
package dataflowbench

object ContextPairDepth2Negative {
  def dfb_source(): String = { // DFB-SOURCE: context-pair-depth2-input
    "tainted"
  }

  def dfb_sink(value: String): Unit = {} // DFB-SINK: context-pair-depth2-sink

  def helper(value: String): String = { // DFB-WITNESS: context-pair-depth2-helper
    value
  }

  def wrapper(value: String): String = { // DFB-WITNESS: context-pair-depth2-wrapper
    helper(value)
  }

  def outerTainted(): String = {
    wrapper(dfb_source())
  }

  def outerClean(): String = {
    wrapper("clean")
  }

  def run(): Unit = {
    val tainted = outerTainted()
    val clean = outerClean()
    dfb_sink(clean)
  }
}

File: cases/taint/cpp/context-pair-depth2-negative/context_pair_depth2_negative.cpp
#include <string>

std::string dfb_source() { // DFB-SOURCE: context-pair-depth2-input
    return "tainted";
}

void dfb_sink(const std::string &value) {} // DFB-SINK: context-pair-depth2-sink

std::string helper(const std::string &value) {
    return value;
}

std::string wrapper(const std::string &value) {
    return helper(value);
}

std::string outer_tainted() {
    return wrapper(dfb_source());
}

std::string outer_clean() {
    return wrapper("clean");
}

void run() {
    std::string tainted = outer_tainted();
    std::string clean = outer_clean();
    dfb_sink(clean);
}
